[WebGPU] Render pipeline may attempt to create a PSO with zero buffers and a vertex shader which requires one or more buffers

mwyrzykowski opened the pull request:
#### 7c012909c61fa9165662458e744e704054dc6674
<pre>
[WebGPU] Render pipeline may attempt to create a PSO with zero buffers and a vertex shader which requires one or more buffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=284608">https://bugs.webkit.org/show_bug.cgi?id=284608</a>
<a href="https://rdar.apple.com/141415294">rdar://141415294</a>

Reviewed by Dan Glastonbury.

The descriptor must have at least as many buffes as the shader requires.

* LayoutTests/fast/webgpu/regression/repro_284608-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_284608.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/RenderPipeline.mm:

Canonical link: <a href="https://commits.webkit.org/287884@main">https://commits.webkit.org/287884@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25eebbae8c4622404487f64d4024ff4d5c05c993

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85714 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32171 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63385 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21149 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/479 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73891 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43683 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30629 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71895 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28622 "Found 1 new test failure: fast/attachment/cocoa/wide-attachment-rendering.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87149 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8415 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71687 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8594 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70922 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17663 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14971 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8376 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13900 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8213 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11733 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10021 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -1545,6 +1545,9 @@ static uint32_t componentsForDataType(MTLDataType dataType)
     }
 
+    if (vertexStageIn && vertexStageIn->size() && !descriptor.vertex.bufferCount)
+        return returnInvalidRenderPipeline(*this, isAsync, @"Vertex descriptor passed zero buffers for stage_in but shader requires expects buffers for stage_in");
```
nit: remove one of `requires` or `expects`

djg reviewed (approved):
:shipit:

mwyrzykowski reviewed (commented):
Added `merge-queue` as test failures were unrelated

mwyrzykowski committed:
[WebGPU] Render pipeline may attempt to create a PSO with zero buffers and a vertex shader which requires one or more buffers
https://bugs.webkit.org/show_bug.cgi?id=284608
rdar://141415294

Reviewed by Dan Glastonbury.

The descriptor must have at least as many buffes as the shader requires.

* LayoutTests/fast/webgpu/regression/repro_284608-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_284608.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/RenderPipeline.mm:

Canonical link: https://commits.webkit.org/287884@main
